File: sdks/python/pmxt/server_manager.py
"""
Server Manager for PMXT Python SDK

This module handles automatic server lifecycle management.
The pattern implemented here is universal and can be replicated in any language SDK.

Universal Pattern:
1. Check if server is running (via lock file + process check)
2. If not running, call pmxt-ensure-server launcher
3. Wait for health check to confirm server is ready
4. Proceed with API calls

This ensures zero-configuration usage across all SDKs.
"""

import os
import json
import time
import subprocess
import shutil
import threading
from pathlib import Path
from typing import List, Optional, Dict, Any
import urllib.request
import urllib.error


class ServerManager:
    """
    Manages the PMXT sidecar server lifecycle.
    
    This class implements the universal server management pattern that
    should be replicated in all language SDKs (Java, C#, Go, etc.)
    """
    
    DEFAULT_PORT = 3847
    HEALTH_CHECK_TIMEOUT = 10  # seconds
    HEALTH_CHECK_INTERVAL = 0.1  # seconds

    # Process-wide coalescing of concurrent ensure_server_running() calls.
    #
    # Each Exchange instance constructs its own ServerManager and each one
    # may call ensure_server_running() from a different thread. Without a
    # shared lock, N concurrent threads all see "no server running", all
    # spawn their own sidecar via pmxt-ensure-server, and the lock file
    # ends up pointing at whichever spawn wrote last. Each Exchange has
    # already captured its own base URL, so most requests end up hitting a
    # sidecar whose access token does NOT match the token read from the
    # lock file, and every request returns 401 Unauthorized.
    #
    # This lock is class-level on purpose — all ServerManager instances in
    # the process share the same sidecar and the same lock file, so they
    # must share the same critical section.
    _ensure_lock = threading.Lock()

    def __init__(self, base_url: str = "http://localhost:3847"):
        """
        Initialize the server manager.
        
        Args:
            base_url: Base URL where server should be running
        """
        self.base_url = base_url
        self.lock_path = Path.home() / '.pmxt' / 'server.lock'
        self._port = self._extract_port_from_url(base_url)
    
    def _extract_port_from_url(self, url: str) -> int:
        """Extract port number from URL."""
        try:
            from urllib.parse import urlparse
            parsed = urlparse(url)
            return parsed.port or self.DEFAULT_PORT
        except:
            return self.DEFAULT_PORT
    
    def ensure_server_running(self) -> None:
        """
        Ensure the PMXT server is running.

        This is the main entry point that SDKs should call.
        It implements the universal pattern:
        1. Check if server is alive
        2. If not, start it via launcher
        3. Wait for health check

        Concurrent calls across all ServerManager instances in the process
        are serialized through a class-level lock so that only one spawn
        attempt happens at a time. See the comment on
        ``ServerManager._ensure_lock`` for why this matters.

        Raises:
            Exception: If server fails to start or become healthy
        """
        with ServerManager._ensure_lock:
            # Step 1: Check if force restart is requested (DEV MODE)
            if os.getenv('PMXT_ALWAYS_RESTART') == '1':
                self._kill_old_server()

            # Step 2: Check if server is already running and matches version.
            # This re-check INSIDE the lock is critical: the thread that won
            # the race will have spawned the sidecar and written the lock
            # file by the time later threads acquire the lock, so they must
            # observe "already running" and return without spawning again.
            if self.is_server_alive():
                if self._is_version_mismatch():
                    self._kill_old_server()
                else:
                    return

            # Step 3: Kill orphan sidecars so the new one can bind the default port
            self._kill_orphan_sidecars()

            # Step 4: Start server via launcher
            self._start_server_via_launcher()

            # Step 5: Wait for health check
            self._wait_for_health()

    def _is_version_mismatch(self) -> bool:
        """Check if running server version matches expected version."""
        server_info = self.get_server_info()
        if not server_info or 'version' not in server_info:
            return True # Old server without version

        # Get expected version
        # 1. Check production path (bundled)
        pkg_path = Path(__file__).parent / '_server' / 'package.json'

        # 2. Check dev path (monorepo)
        if not pkg_path.exists():
            pkg_path = Path(__file__).parent.parent.parent.parent / 'core' / 'package.json'

        if pkg_path.exists():
            try:
                data = json.loads(pkg_path.read_text())
                expected_version = data.get('version')
                server_version = server_info['version']

                if expected_version:
                    # Extract major.minor.patch (ignore prerelease/dev suffixes)
                    def normalize_version(v: str) -> str:
                        """Extract major.minor.patch, ignoring -dev, -b4, etc."""
                        # Remove -dev.xxx or -b4 suffixes
                        base = v.split('-')[0]
                        # Get major.minor.patch
                        parts = base.split('.')[:3]
                        return '.'.join(parts)

                    expected_base = normalize_version(expected_version)
                    server_base = normalize_version(server_version)

                    # Only restart if major.minor.patch differs
                    # This allows 1.0.0 and 1.0.0-b4 to coexist in dev
                    if expected_base != server_base:
                        return True
            except Exception:
                pass

        return False

    def stop(self) -> None:
        """
        Stop the currently running server.

        This reads the lock file to find the process ID and sends a SIGTERM.
        """
        self._kill_old_server()

    def restart(self) -> None:
        """
        Restart the server.

        Stops the current server if running, and starts a fresh one.
        """
        self.stop()
        self.ensure_server_running()

    def start(self) -> None:
        """
        Start the server if it is not already running.

        This is idempotent: if the server is already running and healthy,
        this method returns immediately without restarting.
        """
        self.ensure_server_running()

    def status(self) -> Dict[str, Any]:
        """
        Get a structured snapshot of the sidecar server state.

        Returns a new dict on every call (no shared mutable state). Fields:
            running:        True if the server is alive and responding to /health
            pid:            Process ID from the lock file (None if not running)
            port:           Port the server is listening on (None if not running)
            version:        Server version reported in the lock file (None if absent)
            uptime_seconds: Seconds since the lock file was created (None if absent)
            lock_file:      Absolute path to the lock file
        """
        info = self.get_server_info() or {}
        running = self.is_server_alive()

        timestamp = info.get('timestamp')
        uptime: Optional[float] = None
        if isinstance(timestamp, (int, float)):
            # Lock file timestamps may be epoch seconds or epoch milliseconds.
            now_seconds = time.time()
            ts_seconds = timestamp / 1000.0 if timestamp > 1e12 else float(timestamp)
            delta = now_seconds - ts_seconds
            if delta >= 0:
                uptime = delta

        return {
            'running': running,
            'pid': info.get('pid'),
            'port': info.get('port'),
            'version': info.get('version'),
            'uptime_seconds': uptime,
            'lock_file': str(self.lock_path),
        }

    def health(self) -> bool:
        """
        Check whether the server's /health endpoint is currently responsive.

        Returns:
            True if the server responds with status "ok", False otherwise.
        """
        port = self.get_running_port()
        return self._check_health(port, timeout=2)

    def logs(self, n: int = 50) -> List[str]:
        """
        Return the last `n` lines from the sidecar server log file.

        The launcher writes server stdout/stderr to ~/.pmxt/server.log.
        If the log file does not exist (older launcher, never started, or
        log was cleared), returns an empty list.

        Args:
            n: Maximum number of trailing lines to return (default 50).

        Returns:
            A new list of log line strings (newlines stripped). Returns an
            empty list if no log file is present.
        """
        if n <= 0:
            return []
        log_path = self.lock_path.parent / 'server.log'
        if not log_path.exists():
            return []
        try:
            with log_path.open('r', encoding='utf-8', errors='replace') as f:
                lines = f.readlines()
        except OSError:
            return []
        tail = lines[-n:] if len(lines) > n else list(lines)
        return [line.rstrip('\n') for line in tail]

    def _kill_orphan_sidecars(self) -> None:
        """Kill any orphaned pmxt sidecar processes.

        Orphans accumulate when Python processes exit without stopping their
        detached sidecar.  They occupy ports and can confuse the health check.
        Called before spawning a new sidecar so it can bind the default port.
        """
        import signal as _signal

        try:
            if os.name == 'nt':
                result = subprocess.run(
                    ['wmic', 'process', 'where',
                     "commandline like '%pmxt%bundled.js%'",
                     'get', 'processid'],
                    capture_output=True, text=True, timeout=5,
                )
                for line in result.stdout.strip().splitlines():
                    line = line.strip()
                    if line.isdigit():
                        subprocess.run(
                            ['taskkill', '/PID', line, '/F'],
                            capture_output=True, timeout=5,
                        )
            else:
                result = subprocess.run(
                    ['pgrep', '-f', 'pmxt.*bundled[.]js'],
                    capture_output=True, text=True, timeout=5,
                )
                for line in result.stdout.strip().splitlines():
                    pid = int(line.strip())
                    try:
                        os.kill(pid, _signal.SIGTERM)
                    except (OSError, ProcessLookupError):
                        pass

                if result.stdout.strip():
                    time.sleep(0.5)
        except Exception:
            pass

    def _kill_old_server(self) -> None:
        """Kill the currently running server (Internal)."""
        server_info = self.get_server_info()
        if server_info and 'pid' in server_info:
            pid = server_info['pid']
            try:
                if os.name == 'nt':
                    subprocess.run(
                        ['taskkill', '/PID', str(pid), '/F'],
                        capture_output=True, timeout=5
                    )
                else:
                    import signal
                    os.kill(pid, signal.SIGTERM)
                time.sleep(0.5)
            except Exception:
                pass

            # Verify the process is actually dead; escalate to SIGKILL if not
            if os.name != 'nt':
                try:
                    os.kill(pid, 0)  # raises if dead
                    # Still alive — force kill
                    import signal
                    try:
                        os.kill(pid, signal.SIGKILL)
                        time.sleep(0.2)
                    except Exception:
                        pass
                except (OSError, ProcessLookupError):
                    pass  # Process is dead — good
        self._remove_stale_lock()
    
    def is_server_alive(self) -> bool:
        """
        Check if the server is currently running and healthy.
        
        This implements the universal alive check:
        1. Read lock file
        2. Check if process exists
        3. Optionally verify health endpoint
        
        Returns:
            True if server is running and healthy, False otherwise
        """
        # Check lock file exists
        if not self.lock_path.exists():
            return False
        
        try:
            # Read lock file
            lock_data = json.loads(self.lock_path.read_text())
            pid = lock_data.get('pid')
            port = lock_data.get('port', self.DEFAULT_PORT)
            
            if not pid:
                return False
            
            # Check if process exists (cross-platform)
            if not self._is_process_running(pid):
                # Process doesn't exist, remove stale lock file
                self._remove_stale_lock()
                return False
            
            # Quick health check to verify server is responsive
            try:
                return self._check_health(port, timeout=1)
            except:
                # Process exists but not responding
                return False
                
        except (json.JSONDecodeError, OSError):
            return False
    
    def _is_process_running(self, pid: int) -> bool:
        """
        Check if a process with given PID is running.

        Cross-platform implementation.
        """
        if os.name == 'nt':
            # Windows: signal 0 doesn't work reliably, use tasklist instead
            try:
                result = subprocess.run(
                    ['tasklist', '/FI', f'PID eq {pid}', '/NH'],
                    capture_output=True, text=True, timeout=5
                )
                return str(pid) in result.stdout
            except Exception:
                return False
        else:
            try:
                os.kill(pid, 0)
                return True
            except (OSError, ProcessLookupError):
                return False
    
    def _remove_stale_lock(self) -> None:
        """Remove stale lock file."""
        try:
            self.lock_path.unlink()
        except:
            pass
    
    def _start_server_via_launcher(self) -> None:
        """
        Start the server using the pmxt-ensure-server launcher.
        """
        # 1. Check for bundled server (PRODUCTION - installed via pip)
        launcher_filename = 'pmxt-ensure-server'
        if os.name == "nt": # Check if running Windows
            launcher_filename += ".js"

        bundled_launcher = Path(__file__).parent / '_server' / 'bin' / launcher_filename

        # 2. Check for monorepo structure (DEVELOPMENT)
        current_file = Path(__file__).resolve()
        local_launcher = current_file.parent.parent.parent.parent / 'core' / 'bin' / launcher_filename

        # 3. Check PATH (GLOBAL INSTALL)
        path_launcher = shutil.which(launcher_filename)

        # Priority order: bundled > local dev > PATH
        if bundled_launcher.exists():
            launcher = str(bundled_launcher)
        elif local_launcher.exists():
            launcher = str(local_launcher)
        elif path_launcher:
            launcher = path_launcher
        else:
            raise Exception(
                "pmxt-ensure-server not found.\n"
                "This should have been bundled with the package.\n"
                "Please reinstall: pip install --force-reinstall pmxt\n"
                "Or install the server manually: npm install -g pmxt-core"
            )
        
        # Call the launcher
        try:
            # If it's a JS file, and we are calling it directly, might need node
            cmd = [launcher]
            if launcher.endswith('.js') or not os.access(launcher, os.X_OK):
                cmd = ['node', launcher]

            result = subprocess.run(
                cmd,
                capture_output=True,
                text=True,
                timeout=self.HEALTH_CHECK_TIMEOUT
            )
            
            if result.returncode != 0:
                raise Exception(
                    f"Failed to start server: {result.stderr or result.stdout}"
                )
        except subprocess.TimeoutExpired:
            raise Exception("Server startup timeout")
        except Exception as e:
            raise Exception(f"Failed to start server: {e}")
    
    def _wait_for_health(self) -> None:
        """
        Wait for the server to respond to health checks.

        Reads the port from the lock file on each iteration so that we
        health-check the actual sidecar, not a stale orphan on the default
        port.  Falls back to ``self._port`` when the lock file is absent.
        """
        start_time = time.time()

        while time.time() - start_time < self.HEALTH_CHECK_TIMEOUT:
            try:
                port = self.get_running_port()
                if self._check_health(port):
                    return
            except:
                pass

            time.sleep(self.HEALTH_CHECK_INTERVAL)

        raise Exception(
            f"Server failed to become healthy within {self.HEALTH_CHECK_TIMEOUT}s"
        )
    
    def _check_health(self, port: int, timeout: int = 2) -> bool:
        """
        Check if server is healthy by calling /health endpoint.
        
        Args:
            port: Port to check
            timeout: Request timeout in seconds
            
        Returns:
            True if server responds with 200 OK
        """
        try:
            url = f"http://localhost:{port}/health"
            req = urllib.request.Request(url)
            
            with urllib.request.urlopen(req, timeout=timeout) as response:
                if response.status == 200:
                    data = json.loads(response.read().decode())
                    return data.get('status') == 'ok'
            
            return False
        except (urllib.error.URLError, urllib.error.HTTPError, Exception):
            return False
    
    def get_server_info(self) -> Optional[Dict[str, Any]]:
        """
        Get information about the running server from lock file.
        
        Returns:
            Dictionary with server info (port, pid, timestamp) or None
        """
        if not self.lock_path.exists():
            return None
        
        try:
            return json.loads(self.lock_path.read_text())
        except:
            return None
    
    def get_running_port(self) -> int:
        """
        Get the actual port the server is running on.
        
        This reads the lock file to determine the actual port,
        which may differ from the default if the default port was busy.
        
        Returns:
            Port number the server is running on, or DEFAULT_PORT if unknown
        """
        info = self.get_server_info()
        if info and 'port' in info:
            return info['port']
        return self.DEFAULT_PORT

File: sdks/typescript/pmxt/server-manager.ts
/**
 * Server manager for PMXT TypeScript SDK.
 * 
 * Handles automatic server startup and health checks.
 */

import { DefaultApi, Configuration } from "../generated/src/index.js";
import { readFileSync, existsSync } from "fs";
import { homedir } from "os";
import { join, dirname } from "path";

export interface ServerManagerOptions {
    baseUrl?: string;
    maxRetries?: number;
    retryDelayMs?: number;
}

interface ServerLockInfo {
    port: number;
    pid: number;
    accessToken?: string;
    version?: string;
    timestamp: number;
}

export class ServerManager {
    private baseUrl: string;
    private maxRetries: number;
    private retryDelayMs: number;
    private api: DefaultApi;
    private lockPath: string;
    private static readonly DEFAULT_PORT = 3847;

    // Process-wide coalescing of concurrent ensureServerRunning() calls.
    //
    // Each `Exchange` instance constructs its own ServerManager and each one
    // kicks off ensureServerRunning() from its constructor. Without
    // coalescing, N Exchange instances created in parallel all see "no
    // server running", all spawn their own sidecar via pmxt-ensure-server,
    // and the lock file ends up pointing at whichever spawn wrote last. Each
    // Exchange instance has already captured its own basePath at
    // construction time, so most of them end up talking to a sidecar whose
    // access token does NOT match the token they later read from the lock
    // file — every request returns 401 Unauthorized.
    //
    // The fix is process-wide: when a ensureServerRunning call is in
    // flight, all subsequent callers await the same promise. After the
    // in-flight call settles (success OR failure) the cache is cleared so
    // later callers can re-check the sidecar state (e.g. if it was killed
    // by the user between ticks).
    //
    // This is static on purpose — all ServerManager instances in the
    // process share the same sidecar and the same lock file, so they must
    // share the same in-flight promise.
    private static ensurePromise: Promise<void> | null = null;

    constructor(options: ServerManagerOptions = {}) {
        this.baseUrl = options.baseUrl || `http://localhost:${ServerManager.DEFAULT_PORT}`;
        this.maxRetries = options.maxRetries || 30;
        this.retryDelayMs = options.retryDelayMs || 1000;
        this.lockPath = join(homedir(), '.pmxt', 'server.lock');

        const config = new Configuration({ basePath: this.baseUrl });
        this.api = new DefaultApi(config);
    }

    /**
     * Read server information from lock file.
     */
    private getServerInfo(): ServerLockInfo | null {
        try {
            if (!existsSync(this.lockPath)) {
                return null;
            }
            const content = readFileSync(this.lockPath, 'utf-8');
            return JSON.parse(content) as ServerLockInfo;
        } catch {
            return null;
        }
    }

    /**
     * Get the actual port the server is running on.
     * 
     * This reads the lock file to determine the actual port,
     * which may differ from the default if the default port was busy.
     */
    getRunningPort(): number {
        const info = this.getServerInfo();
        return info?.port || ServerManager.DEFAULT_PORT;
    }

    /**
     * Get the access token from the lock file.
     */
    getAccessToken(): string | undefined {
        const info = this.getServerInfo();
        return info?.accessToken;
    }

    /**
     * Check if the server is running.
     */
    async isServerRunning(): Promise<boolean> {
        // Read lock file to get current port
        const port = this.getRunningPort();

        try {
            // Use native fetch to check health on the actual running port
            // This avoids issues where this.api is configured with the wrong port
            const response = await fetch(`http://localhost:${port}/health`, {
                signal: AbortSignal.timeout(5_000),
            });
            if (response.ok) {
                const data = await response.json();
                return (data as any).status === "ok";
            }
            return false;
        } catch (error) {
            return false;
        }
    }

    /**
     * Wait for the server to be ready.
     * Requires a lock file to be present to avoid falsely matching an unrelated
     * server that may already be running on the default port.
     */
    private async waitForServer(): Promise<void> {
        for (let i = 0; i < this.maxRetries; i++) {
            const info = this.getServerInfo();
            if (info) {
                try {
                    const response = await fetch(`http://localhost:${info.port}/health`, {
                        signal: AbortSignal.timeout(5_000),
                    });
                    if (response.ok) {
                        const data = await response.json() as any;
                        if (data.status === "ok") return;
                    }
                } catch {
                    // Not ready yet
                }
            }
            await new Promise((resolve) => setTimeout(resolve, this.retryDelayMs));
        }
        throw new Error(
            `Server did not start within ${(this.maxRetries * this.retryDelayMs) / 1000}s`
        );
    }

    /**
     * Ensure the server is running, starting it if necessary.
     *
     * Concurrent calls across all ServerManager instances in the process
     * are coalesced onto a single in-flight promise. See the comment on
     * `ServerManager.ensurePromise` for why this matters.
     */
    async ensureServerRunning(): Promise<void> {
        if (ServerManager.ensurePromise) {
            return ServerManager.ensurePromise;
        }
        ServerManager.ensurePromise = this.doEnsureServerRunning().finally(() => {
            ServerManager.ensurePromise = null;
        });
        return ServerManager.ensurePromise;
    }

    private async doEnsureServerRunning(): Promise<void> {
        // Check for force restart
        if (process.env.PMXT_ALWAYS_RESTART === '1') {
            await this.killOldServer();
        }

        // Check if already running and version matches
        if (await this.isServerRunning()) {
            if (await this.isVersionMismatch()) {
                await this.killOldServer();
            } else {
                return;
            }
        }

        // Locate pmxt-ensure-server
        const isWindows = process.platform === 'win32';
        const launcherName = isWindows ? 'pmxt-ensure-server.js' : 'pmxt-ensure-server';
        let launcherPath = launcherName; // Default to PATH

        try {
            // Try to resolve from pmxt-core dependency
            // For CommonJS build (which is primary), we can use require directly
            // For ESM build, this will be transpiled appropriately
            const corePackageJson = require.resolve('pmxt-core/package.json');
            const coreDir = dirname(corePackageJson);
            const binPath = join(coreDir, 'bin', launcherName);

            if (existsSync(binPath)) {
                launcherPath = binPath;
            }
        } catch (error: unknown) {
            if (process.env.PMXT_LOG_LEVEL === 'debug') {
                const msg = error instanceof Error ? error.message : String(error);
                process.stderr.write(`[pmxt] Binary path resolution failed, falling back to PATH: ${msg}\n`);
            }
        }

        // Try to start the server using pmxt-ensure-server
        const { spawn } = await import("child_process");

        // On Windows, .js scripts must be run via node explicitly
        const spawnCmd = launcherPath.endsWith('.js') ? 'node' : launcherPath;
        const spawnArgs = launcherPath.endsWith('.js') ? [launcherPath] : [];

        try {
            const proc = spawn(spawnCmd, spawnArgs, {
                detached: true,
                stdio: "ignore",
            });
            proc.unref();

            // Wait for server to be ready
            await this.waitForServer();
        } catch (error) {
            throw new Error(
                `Failed to start PMXT server: ${error}\n\n` +
                `Please ensure 'pmxt-core' is installed: npm install -g pmxt-core\n` +
                `Or start the server manually: pmxt-server`
            );
        }
    }

    private async isVersionMismatch(): Promise<boolean> {
        const info = this.getServerInfo();
        if (!info || !info.version) {
            return true; // Old server without version
        }

        try {
            // 1. Try to find package.json relative to the installed location (Production)
            let corePackageJsonPath: string | undefined;
            try {
                corePackageJsonPath = require.resolve('pmxt-core/package.json');
            } catch {
                // 2. Try dev path (Monorepo)
                const devPath = join(dirname(__dirname), '../../core/package.json');
                if (existsSync(devPath)) {
                    corePackageJsonPath = devPath;
                }
            }

            if (corePackageJsonPath && existsSync(corePackageJsonPath)) {
                const content = readFileSync(corePackageJsonPath, 'utf-8');
                const pkg = JSON.parse(content);
                // Check if running version starts with package version
                // (Server version might have extra hash in dev mode)
                if (pkg.version && !info.version.startsWith(pkg.version)) {
                    return true;
                }
            }
        } catch {
            // Ignore errors
        }
        return false;
    }

    /**
     * Stop the currently running server.
     */
    async stop(): Promise<void> {
        await this.killOldServer();
    }

    /**
     * Restart the server.
     */
    async restart(): Promise<void> {
        await this.stop();
        await this.ensureServerRunning();
    }

    /**
     * Start the server if it is not already running.
     *
     * Idempotent: if the server is already running and healthy this returns
     * immediately without restarting.
     */
    async start(): Promise<void> {
        await this.ensureServerRunning();
    }

    /**
     * Get a structured snapshot of the sidecar server state.
     *
     * Returns a fresh object on every call (no shared mutable state).
     */
    async status(): Promise<{
        running: boolean;
        pid: number | null;
        port: number | null;
        version: string | null;
        uptimeSeconds: number | null;
        lockFile: string;
    }> {
        const info = this.getServerInfo();
        const running = await this.isServerRunning();

        let uptimeSeconds: number | null = null;
        if (info && typeof info.timestamp === "number") {
            const nowSeconds = Date.now() / 1000;
            const tsSeconds = info.timestamp > 1e12 ? info.timestamp / 1000 : info.timestamp;
            const delta = nowSeconds - tsSeconds;
            if (delta >= 0) uptimeSeconds = delta;
        }

        return {
            running,
            pid: info?.pid ?? null,
            port: info?.port ?? null,
            version: info?.version ?? null,
            uptimeSeconds,
            lockFile: this.lockPath,
        };
    }

    /**
     * Check whether the server's /health endpoint is currently responsive.
     */
    async health(): Promise<boolean> {
        return this.isServerRunning();
    }

    /**
     * Return the last `n` lines from the sidecar server log file.
     *
     * The launcher writes server stdout/stderr to ~/.pmxt/server.log.
     * Returns an empty array if no log file is present.
     */
    logs(n: number = 50): string[] {
        if (n <= 0) return [];
        const logPath = join(dirname(this.lockPath), "server.log");
        try {
            if (!existsSync(logPath)) return [];
            const content = readFileSync(logPath, "utf-8");
            const lines = content.split(/\r?\n/);
            // split on a trailing newline produces an empty final element; drop it
            if (lines.length > 0 && lines[lines.length - 1] === "") lines.pop();
            return lines.length > n ? lines.slice(lines.length - n) : lines;
        } catch {
            return [];
        }
    }

    private async killOldServer(): Promise<void> {
        const info = this.getServerInfo();
        if (info && info.pid) {
            try {
                process.kill(info.pid, 'SIGTERM');
                await new Promise(resolve => setTimeout(resolve, 500));
            } catch {
                // Process already dead — fall through to lock cleanup
            }

            // Verify the process is actually dead; escalate to SIGKILL if not
            try {
                process.kill(info.pid, 0); // throws if dead
                // Still alive — force kill
                try {
                    process.kill(info.pid, 'SIGKILL');
                    await new Promise(resolve => setTimeout(resolve, 200));
                } catch {
                    // Ignore — SIGKILL may race with natural exit
                }
            } catch {
                // Process is dead — good
            }
        }
        // Remove lock file (best effort)
        try {
            const { unlinkSync } = await import('fs');
            unlinkSync(this.lockPath);
        } catch { }
    }
}
